constraints.pro

```diff
@@ -131,100 +149,116 @@ gen_enforced_field(WorkspaceCwd, 'description', DescriptionWithoutTrailingPeriod
-% "keywords" must be the same across all workspace packages
-% (and must be unset for the root).
+% All published packages must have the same set of NPM keywords.
 gen_enforced_field(WorkspaceCwd, 'keywords', ['MetaMask', 'Ethereum']) :-
   \+ workspace_field(WorkspaceCwd, 'private', true).
+% Non-published packages do not have any NPM keywords.
 gen_enforced_field(WorkspaceCwd, 'keywords', null) :-
   workspace_field(WorkspaceCwd, 'private', true).
 
-% "homepage" must match the name of the package (based on the workspace
-% directory name) across all workspace packages (and must be unset for the
-% root).
+% The homepage of a published package must match its name (which is in turn
+% based on its workspace directory name).
 gen_enforced_field(WorkspaceCwd, 'homepage', CorrectHomepageUrl) :-
   \+ workspace_field(WorkspaceCwd, 'private', true),
   workspace_basename(WorkspaceCwd, WorkspaceBasename),
-  atomic_list_concat(['https://github.com/MetaMask/controllers/tree/main/packages/', WorkspaceBasename, '#readme'], CorrectHomepageUrl).
+  workspace_field(WorkspaceCwd, 'repository.url', RepoUrl),
+  repo_name(RepoUrl, RepoName),
+  atomic_list_concat(['https://github.com/MetaMask/', RepoName, '/tree/main/packages/', WorkspaceBasename, '#readme'], CorrectHomepageUrl).
+% Non-published packages do not have a homepage.
 gen_enforced_field(WorkspaceCwd, 'homepage', null) :-
   workspace_field(WorkspaceCwd, 'private', true).
 
-% "repository.type" must be "git" for all packages.
-gen_enforced_field(WorkspaceCwd, 'repository.type', 'git').
-
-% "repository.url" must be "https://github.com/MetaMask/controllers.git" for all
-% packages.
-gen_enforced_field(WorkspaceCwd, 'repository.url', 'https://github.com/MetaMask/controllers.git').
-
-% "bugs.url" must be "https://github.com/MetaMask/controllers/issues" for all
-% workspace packages.
-gen_enforced_field(WorkspaceCwd, 'bugs.url', 'https://github.com/MetaMask/controllers/issues') :-
-  \+ workspace_field(WorkspaceCwd, 'private', true).
-
-% "bugs" must unset for the root.
+% The bugs URL of a published package must point to the Issues page for the
+% repository.
+gen_enforced_field(WorkspaceCwd, 'bugs.url', CorrectBugsUrl) :-
+  \+ workspace_field(WorkspaceCwd, 'private', true),
+  workspace_field(WorkspaceCwd, 'repository.url', RepoUrl),
+  repo_name(RepoUrl, RepoName),
+  atomic_list_concat(['https://github.com/MetaMask/', RepoName, '/issues'], CorrectBugsUrl).
+% Non-published packages must not have a bugs section.
 gen_enforced_field(WorkspaceCwd, 'bugs', null) :-
   workspace_field(WorkspaceCwd, 'private', true).
 
-% "license" must be "MIT" for all workspace packages and unset for the root.
+% All packages must specify Git as the repository type.
+gen_enforced_field(WorkspaceCwd, 'repository.type', 'git').
+
+% All packages must match the URL of a repo within the MetaMask organization.
+gen_enforced_field(WorkspaceCwd, 'repository.url', 'https://github.com/MetaMask/<insert repo name here>.git') :-
```

Comment:
Hmm. I found this confusing - what is this `'https://github.com/MetaMask/<insert repo name here>.git'` value being used for?

The docs for `gen_enforced_field` suggest that it's going to enforce this as being the real value set here, but clearly that's not happening.

Reply:
This is being used to print a suggestion to the user when the URL is not in the correct format. The rule here says "The repository URL should be 'https://github.com/MetaMask/<insert repo name here>.git' when a repo name cannot be identified within the current version of the URL". If you change `repository.url` in the root `package.json` to, say, "foo", then you will this message:

```
➤ YN0038: @metamask/controllers-monorepo must have a field repository.url set to "https://github.com/MetaMask/<insert repo name here>.git", but is set to "foo" instead
```

Reply:
Got it, makes sense